pretyping/heads.ml

```diff
@@ -124,22 +121,16 @@ let kind_of_head env t =
 (* FIXME: maybe change interface here *)
 let compute_head = function
 | EvalConstRef cst ->
-   let env = Global.env() in
-   let cb = Environ.lookup_constant cst env in
-   let is_Def = function Declarations.Def _ -> true | _ -> false in
-   let body =
-     if not (Recordops.is_primitive_projection cst) && is_Def cb.Declarations.const_body
-     then Global.body_of_constant cst else None
-   in
-     (match body with
-     | None -> RigidHead (RigidParameter cst)
-     | Some (c, _) -> kind_of_head env c)
+  let env = Global.env() in
+  let body = Environ.constant_opt_value_in env (cst,Univ.Instance.empty) in
+  (match body with
+   | None -> RigidHead (RigidParameter cst)
+   | Some c -> kind_of_head env c)
```

Comment:
IIUC. Now we will return RigidHead RigidOther instead of RigidHead (RigidParameter cst) if given a cst wich is a primitive projection?

I'm not sure this code is correct. It would seem more meaningful to follow the Case case in kind_of_head for Proj(p,c). And say that the rigidity of Proj(p,c) is the rigidity of c. @herbelin does that sound right?

Reply:
>I'm not sure this code is correct.

I am sure that it is no more incorrect than before this PR though ;)

Reply:
In the `kind_of_head` for `Proj`, above, what about first testing the head of `c`? If it is a `ConstructorHead`, as in the `Case` case, we renounce to do the contraction step and returns `NotImmediatelyComputableHead`. Otherwise, we return the head of `c`.

And for `compute_head`, using the code of @SkySkimmer.

Reply:
@SkySkimmer well, it is maybe not more incorrect but it returns a different result, that's all I'm saying. We can leave @herbelin 's proposal which sounds right to me for another PR if you prefer.